vm/ubpf_instruction_valid.c

```diff
@@ -0,0 +1,1010 @@
+        .immediate_upper_bound = INT32_MAX,
+    },
+    {
+        .opcode = EBPF_OP_LSH_REG,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .source_lower_bound = BPF_REG_0,
+        .source_upper_bound = BPF_REG_10,
+    },
+    {
+        .opcode = EBPF_OP_RSH_IMM,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .immediate_lower_bound = INT32_MIN,
+        .immediate_upper_bound = INT32_MAX,
+    },
+    {
+        .opcode = EBPF_OP_RSH_REG,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .source_lower_bound = BPF_REG_0,
+        .source_upper_bound = BPF_REG_10,
+    },
+    {
+        .opcode = EBPF_OP_NEG,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+    },
+    {
+        .opcode = EBPF_OP_MOD_IMM,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .immediate_lower_bound = INT32_MIN,
+        .immediate_upper_bound = INT32_MAX,
+    },
+    {
+        .opcode = EBPF_OP_MOD_REG,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .source_lower_bound = BPF_REG_0,
+        .source_upper_bound = BPF_REG_10,
+    },
+    {
+        .opcode = EBPF_OP_XOR_IMM,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .immediate_lower_bound = INT32_MIN,
+        .immediate_upper_bound = INT32_MAX,
+    },
+    {
+        .opcode = EBPF_OP_XOR_REG,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .source_lower_bound = BPF_REG_0,
+        .source_upper_bound = BPF_REG_10,
+    },
+    {
+        .opcode = EBPF_OP_MOV_IMM,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .immediate_lower_bound = INT32_MIN,
+        .immediate_upper_bound = INT32_MAX,
+    },
+    {
+        .opcode = EBPF_OP_MOV_REG,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .source_lower_bound = BPF_REG_0,
+        .source_upper_bound = BPF_REG_10,
+    },
+    {
+        .opcode = EBPF_OP_ARSH_IMM,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .immediate_lower_bound = INT32_MIN,
+        .immediate_upper_bound = INT32_MAX,
+    },
+    {
+        .opcode = EBPF_OP_ARSH_REG,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .source_lower_bound = BPF_REG_0,
+        .source_upper_bound = BPF_REG_10,
+    },
+    {
+        .opcode = EBPF_OP_LE,
+        .destination_lower_bound = BPF_REG_0,
+        .destination_upper_bound = BPF_REG_9,
+        .immediate_lower_bound = 0,
+        .immediate_upper_bound = 64,
```

Comment:
This does seem to handle the upper limit, but according to the spec, not every value between `0` and `64` is valid. Only the values 16, 32, and 64 are allowed (https://www.ietf.org/archive/id/draft-ietf-bpf-isa-04.html#section-4.2-5). Does the filter technique here handle that?